File: core/infra/src/test/java/com/beat/infra/external/social/kakao/KakaoSocialLoginAdapterTest.java
package com.beat.infra.external.social.kakao;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertSame;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

import java.net.SocketTimeoutException;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.test.util.ReflectionTestUtils;

import com.beat.contracts.auth.social.SocialLoginFailure;
import com.beat.contracts.auth.social.SocialLoginRequest;
import com.beat.contracts.auth.social.SocialLoginType;
import com.beat.infra.external.social.kakao.client.KakaoApiClient;
import com.beat.infra.external.social.kakao.client.KakaoAuthApiClient;
import com.beat.infra.external.social.kakao.response.KakaoAccessTokenResponse;

import feign.FeignException;
import feign.RetryableException;

@ExtendWith(MockitoExtension.class)
class KakaoSocialLoginAdapterTest {

	@Mock
	private KakaoApiClient kakaoApiClient;

	@Mock
	private KakaoAuthApiClient kakaoAuthApiClient;

	private KakaoSocialLoginAdapter adapter;

	@BeforeEach
	void setUp() {
		adapter = new KakaoSocialLoginAdapter(kakaoApiClient, kakaoAuthApiClient);
		ReflectionTestUtils.setField(adapter, "clientId", "client-id");
		ReflectionTestUtils.setField(adapter, "redirectUri", "redirect-uri");
	}

	@Test
	void providerServerErrorIsNotMisclassifiedAsAuthenticationFailure() {
		FeignException providerException = mock(FeignException.class);
		when(providerException.status()).thenReturn(503);
		when(kakaoAuthApiClient.getOAuth2AccessToken(anyString(), anyString(), anyString(), anyString()))
			.thenThrow(providerException);

		SocialLoginFailure failure = assertThrows(SocialLoginFailure.class, () -> adapter.login(kakaoRequest()));

		assertEquals(SocialLoginFailure.Reason.PROVIDER_UNAVAILABLE, failure.getReason());
		assertSame(providerException, failure.getCause());
	}

	@Test
	void providerRateLimitIsUnavailableInsteadOfAuthenticationFailure() {
		FeignException rateLimitException = mock(FeignException.class);
		when(rateLimitException.status()).thenReturn(429);
		when(kakaoAuthApiClient.getOAuth2AccessToken(anyString(), anyString(), anyString(), anyString()))
			.thenThrow(rateLimitException);

		SocialLoginFailure failure = assertThrows(SocialLoginFailure.class, () -> adapter.login(kakaoRequest()));

		assertEquals(SocialLoginFailure.Reason.PROVIDER_UNAVAILABLE, failure.getReason());
		assertSame(rateLimitException, failure.getCause());
	}

	@Test
	void providerConfigurationErrorIsBadGatewayInsteadOfAuthenticationFailure() {
		FeignException providerException = mock(FeignException.class);
		when(providerException.status()).thenReturn(403);
		when(kakaoAuthApiClient.getOAuth2AccessToken(anyString(), anyString(), anyString(), anyString()))
			.thenThrow(providerException);

		SocialLoginFailure failure = assertThrows(SocialLoginFailure.class, () -> adapter.login(kakaoRequest()));

		assertEquals(SocialLoginFailure.Reason.PROVIDER_FAILURE, failure.getReason());
		assertSame(providerException, failure.getCause());
	}

	@Test
	void rejectedAuthorizationCodeIsAuthenticationFailure() {
		FeignException authenticationException = mock(FeignException.class);
		when(authenticationException.status()).thenReturn(400);
		when(authenticationException.contentUTF8())
			.thenReturn("{\"error\":\"invalid_grant\",\"error_code\":\"KOE320\"}");
		when(kakaoAuthApiClient.getOAuth2AccessToken(anyString(), anyString(), anyString(), anyString()))
			.thenThrow(authenticationException);

		SocialLoginFailure failure = assertThrows(SocialLoginFailure.class, () -> adapter.login(kakaoRequest()));

		assertEquals(SocialLoginFailure.Reason.AUTHENTICATION_FAILED, failure.getReason());
		assertSame(authenticationException, failure.getCause());
	}

	@Test
	void providerConfigurationBadRequestIsNotAuthenticationFailure() {
		FeignException configurationException = mock(FeignException.class);
		when(configurationException.status()).thenReturn(400);
		when(configurationException.contentUTF8())
			.thenReturn("{\"error\":\"invalid_client\",\"error_code\":\"KOE101\"}");
		when(kakaoAuthApiClient.getOAuth2AccessToken(anyString(), anyString(), anyString(), anyString()))
			.thenThrow(configurationException);

		SocialLoginFailure failure = assertThrows(SocialLoginFailure.class, () -> adapter.login(kakaoRequest()));

		assertEquals(SocialLoginFailure.Reason.PROVIDER_FAILURE, failure.getReason());
		assertSame(configurationException, failure.getCause());
	}

	@Test
	void rejectedProviderAccessTokenIsProviderFailure() {
		FeignException authenticationException = mock(FeignException.class);
		when(authenticationException.status()).thenReturn(401);
		when(kakaoAuthApiClient.getOAuth2AccessToken(anyString(), anyString(), anyString(), anyString()))
			.thenReturn(successfulTokenResponse());
		when(kakaoApiClient.getUserInformation("Bearer access-token")).thenThrow(authenticationException);

		SocialLoginFailure failure = assertThrows(SocialLoginFailure.class, () -> adapter.login(kakaoRequest()));

		assertEquals(SocialLoginFailure.Reason.PROVIDER_FAILURE, failure.getReason());
		assertSame(authenticationException, failure.getCause());
	}

	@Test
	void userInfoProviderErrorIsBadGatewayInsteadOfAuthenticationFailure() {
		FeignException providerException = mock(FeignException.class);
		when(providerException.status()).thenReturn(403);
		when(kakaoAuthApiClient.getOAuth2AccessToken(anyString(), anyString(), anyString(), anyString()))
			.thenReturn(successfulTokenResponse());
		when(kakaoApiClient.getUserInformation("Bearer access-token")).thenThrow(providerException);

		SocialLoginFailure failure = assertThrows(SocialLoginFailure.class, () -> adapter.login(kakaoRequest()));

		assertEquals(SocialLoginFailure.Reason.PROVIDER_FAILURE, failure.getReason());
		assertSame(providerException, failure.getCause());
	}

	@Test
	void retryableTimeoutIsClassifiedSeparately() {
		RetryableException timeoutException = mock(RetryableException.class);
		when(timeoutException.getCause()).thenReturn(new SocketTimeoutException("timeout"));
		when(kakaoAuthApiClient.getOAuth2AccessToken(anyString(), anyString(), anyString(), anyString()))
			.thenThrow(timeoutException);

		SocialLoginFailure failure = assertThrows(SocialLoginFailure.class, () -> adapter.login(kakaoRequest()));

		assertEquals(SocialLoginFailure.Reason.PROVIDER_TIMEOUT, failure.getReason());
		assertSame(timeoutException, failure.getCause());
	}

	@Test
	void missingTokenResponseIsProviderFailure() {
		when(kakaoAuthApiClient.getOAuth2AccessToken(anyString(), anyString(), anyString(), anyString()))
			.thenReturn(null);

		SocialLoginFailure failure = assertThrows(SocialLoginFailure.class, () -> adapter.login(kakaoRequest()));

		assertEquals(SocialLoginFailure.Reason.PROVIDER_FAILURE, failure.getReason());
	}

	private static SocialLoginRequest kakaoRequest() {
		return new SocialLoginRequest("authorization-code", SocialLoginType.KAKAO);
	}

	private static KakaoAccessTokenResponse successfulTokenResponse() {
		return new KakaoAccessTokenResponse("Bearer", "access-token", 3600, "refresh-token", 7200);
	}
}

File: core/infra/src/test/kotlin/com/beat/infra/external/social/kakao/KakaoSocialLoginAdapterTest.kt
package com.beat.infra.external.social.kakao

import com.beat.application.frontoffice.member.command.SocialLoginFailure
import com.beat.application.frontoffice.member.command.SocialLoginRequest
import com.beat.application.frontoffice.member.command.SocialLoginType
import com.beat.infra.external.social.kakao.client.KakaoApiClient
import com.beat.infra.external.social.kakao.client.KakaoAuthApiClient
import com.beat.infra.external.social.kakao.response.KakaoAccessTokenResponse
import com.beat.infra.external.social.kakao.response.KakaoAccount
import com.beat.infra.external.social.kakao.response.KakaoUserProfile
import com.beat.infra.external.social.kakao.response.KakaoUserResponse
import feign.FeignException
import feign.RetryableException
import io.kotest.assertions.throwables.shouldThrow
import io.kotest.core.spec.style.FunSpec
import io.kotest.matchers.shouldBe
import java.net.SocketTimeoutException
import org.mockito.ArgumentMatchers.anyString
import org.mockito.Mockito.mock
import org.mockito.Mockito.`when` as given
import org.springframework.test.util.ReflectionTestUtils

class KakaoSocialLoginAdapterTest : FunSpec({

    var kakaoApiClient = mock(KakaoApiClient::class.java)
    var kakaoAuthApiClient = mock(KakaoAuthApiClient::class.java)
    var adapter = KakaoSocialLoginAdapter(kakaoApiClient, kakaoAuthApiClient)

    beforeTest {
        kakaoApiClient = mock(KakaoApiClient::class.java)
        kakaoAuthApiClient = mock(KakaoAuthApiClient::class.java)
        adapter = KakaoSocialLoginAdapter(kakaoApiClient, kakaoAuthApiClient)
        ReflectionTestUtils.setField(adapter, "clientId", "client-id")
        ReflectionTestUtils.setField(adapter, "redirectUri", "redirect-uri")
    }

    test("provider server error is not misclassified as authentication failure") {
        val providerException = mock(FeignException::class.java)
        given(providerException.status()).thenReturn(503)
        given(kakaoAuthApiClient.getOAuth2AccessToken(anyString(), anyString(), anyString(), anyString()))
            .thenThrow(providerException)

        val failure = shouldThrow<SocialLoginFailure> { adapter.login(kakaoRequest()) }

        failure.reason shouldBe SocialLoginFailure.Reason.PROVIDER_UNAVAILABLE
        (failure.cause === providerException) shouldBe true
    }

    test("provider rate limit is unavailable instead of authentication failure") {
        val rateLimitException = mock(FeignException::class.java)
        given(rateLimitException.status()).thenReturn(429)
        given(kakaoAuthApiClient.getOAuth2AccessToken(anyString(), anyString(), anyString(), anyString()))
            .thenThrow(rateLimitException)

        val failure = shouldThrow<SocialLoginFailure> { adapter.login(kakaoRequest()) }

        failure.reason shouldBe SocialLoginFailure.Reason.PROVIDER_UNAVAILABLE
        (failure.cause === rateLimitException) shouldBe true
    }

    test("provider configuration error is provider failure") {
        val providerException = mock(FeignException::class.java)
        given(providerException.status()).thenReturn(403)
        given(kakaoAuthApiClient.getOAuth2AccessToken(anyString(), anyString(), anyString(), anyString()))
            .thenThrow(providerException)

        val failure = shouldThrow<SocialLoginFailure> { adapter.login(kakaoRequest()) }

        failure.reason shouldBe SocialLoginFailure.Reason.PROVIDER_FAILURE
        (failure.cause === providerException) shouldBe true
    }

    test("rejected authorization code is authentication failure") {
        val authenticationException = mock(FeignException::class.java)
        given(authenticationException.status()).thenReturn(400)
        given(authenticationException.contentUTF8())
            .thenReturn("{\"error\":\"invalid_grant\",\"error_code\":\"KOE320\"}")
        given(kakaoAuthApiClient.getOAuth2AccessToken(anyString(), anyString(), anyString(), anyString()))
            .thenThrow(authenticationException)

        val failure = shouldThrow<SocialLoginFailure> { adapter.login(kakaoRequest()) }

        failure.reason shouldBe SocialLoginFailure.Reason.AUTHENTICATION_FAILED
        (failure.cause === authenticationException) shouldBe true
    }

    test("provider configuration bad request is not authentication failure") {
        val configurationException = mock(FeignException::class.java)
        given(configurationException.status()).thenReturn(400)
        given(configurationException.contentUTF8())
            .thenReturn("{\"error\":\"invalid_client\",\"error_code\":\"KOE101\"}")
        given(kakaoAuthApiClient.getOAuth2AccessToken(anyString(), anyString(), anyString(), anyString()))
            .thenThrow(configurationException)

        val failure = shouldThrow<SocialLoginFailure> { adapter.login(kakaoRequest()) }

        failure.reason shouldBe SocialLoginFailure.Reason.PROVIDER_FAILURE
        (failure.cause === configurationException) shouldBe true
    }

    test("rejected provider access token is provider failure") {
        val authenticationException = mock(FeignException::class.java)
        given(authenticationException.status()).thenReturn(401)
        given(kakaoAuthApiClient.getOAuth2AccessToken(anyString(), anyString(), anyString(), anyString()))
            .thenReturn(successfulTokenResponse())
        given(kakaoApiClient.getUserInformation("Bearer access-token")).thenThrow(authenticationException)

        val failure = shouldThrow<SocialLoginFailure> { adapter.login(kakaoRequest()) }

        failure.reason shouldBe SocialLoginFailure.Reason.PROVIDER_FAILURE
        (failure.cause === authenticationException) shouldBe true
    }

    test("user info provider error is provider failure") {
        val providerException = mock(FeignException::class.java)
        given(providerException.status()).thenReturn(403)
        given(kakaoAuthApiClient.getOAuth2AccessToken(anyString(), anyString(), anyString(), anyString()))
            .thenReturn(successfulTokenResponse())
        given(kakaoApiClient.getUserInformation("Bearer access-token")).thenThrow(providerException)

        val failure = shouldThrow<SocialLoginFailure> { adapter.login(kakaoRequest()) }

        failure.reason shouldBe SocialLoginFailure.Reason.PROVIDER_FAILURE
        (failure.cause === providerException) shouldBe true
    }

    test("retryable timeout is classified separately") {
        val timeoutException = mock(RetryableException::class.java)
        given(timeoutException.cause).thenReturn(SocketTimeoutException("timeout"))
        given(kakaoAuthApiClient.getOAuth2AccessToken(anyString(), anyString(), anyString(), anyString()))
            .thenThrow(timeoutException)

        val failure = shouldThrow<SocialLoginFailure> { adapter.login(kakaoRequest()) }

        failure.reason shouldBe SocialLoginFailure.Reason.PROVIDER_TIMEOUT
        (failure.cause === timeoutException) shouldBe true
    }

    test("missing token response is provider failure") {
        given(kakaoAuthApiClient.getOAuth2AccessToken(anyString(), anyString(), anyString(), anyString()))
            .thenReturn(null)

        val failure = shouldThrow<SocialLoginFailure> { adapter.login(kakaoRequest()) }

        failure.reason shouldBe SocialLoginFailure.Reason.PROVIDER_FAILURE
    }

    test("successful response maps member profile") {
        given(kakaoAuthApiClient.getOAuth2AccessToken(anyString(), anyString(), anyString(), anyString()))
            .thenReturn(successfulTokenResponse())
        given(kakaoApiClient.getUserInformation("Bearer access-token")).thenReturn(
            KakaoUserResponse(
                id = 123L,
                connectedAt = null,
                kakaoAccount = KakaoAccount(
                    profileNicknameNeedsAgreement = false,
                    profile = KakaoUserProfile("nickname", null, null, false),
                    emailNeedsAgreement = false,
                    emailValid = true,
                    emailVerified = true,
                    email = "member@example.com",
                ),
            ),
        )

        val profile = adapter.login(kakaoRequest())

        profile.socialId shouldBe 123L
        profile.nickname shouldBe "nickname"
        profile.email shouldBe "member@example.com"
    }

})

private fun kakaoRequest(): SocialLoginRequest =
    SocialLoginRequest("authorization-code", SocialLoginType.KAKAO)

private fun successfulTokenResponse(): KakaoAccessTokenResponse =
    KakaoAccessTokenResponse("Bearer", "access-token", 3600, "refresh-token", 7200)
